python/delta-kernel-python/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+
+#[pymethods]
+impl Snapshot {
+    fn version(&self) -> delta_kernel::Version {
+        self.0.version()
+    }
+}
+
+#[pyclass]
+struct ScanBuilder(Option<delta_kernel::scan::ScanBuilder>);
+
+#[pymethods]
+impl ScanBuilder {
+    #[new]
+    fn new(snapshot: &Snapshot) -> ScanBuilder {
+        let sb = delta_kernel::scan::ScanBuilder::new(snapshot.0.clone());
+        ScanBuilder(Some(sb))
+    }
+
+    fn build(&mut self) -> DeltaPyResult<Scan> {
+        let scan = self.0.take().unwrap().build()?;
+        Ok(Scan(scan))
+    }
+}
+
+#[pyclass]
+struct Scan(delta_kernel::scan::Scan);
+
+#[pymethods]
+impl Scan {
+    fn execute(
+        &self,
+        engine_interface: &PythonInterface,
+    ) -> DeltaPyResult<PyArrowType<Box<dyn RecordBatchReader + Send>>> {
+        let mut results = self.0.execute(engine_interface.0.as_ref())?;
+        match results.pop() {
+            Some(last) => {
+                // TODO(nick): This is a terrible way to have to get the schema
+                let rb: RecordBatch = last
+                    .raw_data?
+                    .into_any()
+                    .downcast::<ArrowEngineData>()
+                    .map_err(|_| {
+                        delta_kernel::Error::engine_data_type("Couldn't cast to ArrowEngineData")
+                    })?
+                    .into();
+                let schema: SchemaRef = rb.schema();
+                results.push(ScanResult {
+                    raw_data: Ok(Box::new(ArrowEngineData::new(rb))),
+                    mask: last.mask,
+                });
+                let record_batch_iter = RecordBatchIterator::new(
+                    results.into_iter().map(|res| {
+                        let data = res
+                            .raw_data
+                            .map_err(|e| ArrowError::from_external_error(Box::new(e)))?;
+                        let record_batch: RecordBatch = data
+                            .into_any()
+                            .downcast::<ArrowEngineData>()
+                            .map_err(|_| {
+                                ArrowError::CastError("Couldn't cast to ArrowEngineData".to_string())
+                            })?
+                            .into();
+                        if let Some(mask) = res.mask {
+                            let filtered_batch = filter_record_batch(&record_batch, &mask.into())?;
+                            Ok(filtered_batch)
+                        } else {
+                            Ok(record_batch)
+                        }
+                    }),
+                    schema,
+                );
+                Ok(PyArrowType(Box::new(record_batch_iter)))
+            }
+            None => {
+                // no results, return empty iterator
+                Ok(PyArrowType(Box::new(RecordBatchIterator::new(
+                    vec![],
+                    Arc::new(Schema::empty()),
+                ))))
+            }
+        }
+    }
```

Comment:
```suggestion
         let result_schema: SchemaRef =
            Arc::new(self.0.schema().as_ref().try_into().map_err(|e| {
                delta_kernel::Error::Generic(format!("Could not get result schema: {e}"))
            })?);
        let results = self.0.execute(engine_interface.0.as_ref())?;
        let record_batch_iter = RecordBatchIterator::new(
            results.into_iter().map(|res| {
                let data = res
                    .raw_data
                    .map_err(|e| ArrowError::from_external_error(Box::new(e)))?;
                let record_batch: RecordBatch = data
                    .into_any()
                    .downcast::<ArrowEngineData>()
                    .map_err(|_| {
                        ArrowError::CastError("Couldn't cast to ArrowEngineData".to_string())
                    })?
                    .into();
                if let Some(mut mask) = res.mask {
                    let extra_rows = record_batch.num_rows() - mask.len();
                    if extra_rows > 0 {
                        // we need to extend the mask here in case it's too short
                        mask.extend(std::iter::repeat(true).take(extra_rows));
                    }
                    let filtered_batch = filter_record_batch(&record_batch, &mask.into())?;
                    Ok(filtered_batch)
                } else {
                    Ok(record_batch)
                }
            }),
            result_schema,
        );
        Ok(PyArrowType(Box::new(record_batch_iter)))
```